amplipi/streams.py

```diff
@@ -1321,6 +1323,216 @@ def info(self) -> models.SourceInfo:
+    self.change_song(self.song_index - 1)
+
+  def change_song(self, new_song_id):
+    self.song_index = new_song_id
+    self.playing = self.song_list[self.song_index]
+    last_src = self.src
+    self.restart()
+
+    f = open(self.command_file_path, 'w')
+    f.write('play')
+    f.close()
+    self._ended_timeout = datetime.datetime.now() + datetime.timedelta(seconds=3)
+    self.ended = False
+
+  def send_cmd(self, cmd):
+    if cmd in self.supported_cmds:
+      if cmd == 'stop':
+        if self._is_running():
+          self.proc.kill()
+          if self.bkg_thread:
+            self.bkg_thread.join()
+        self.proc = None
+      if self.command_file_path is not None:
+        if cmd == 'pause':
+          f = open(self.command_file_path, 'w')
+          f.write('pause')
+          f.close()
+          self.state = 'paused'
+
+        if cmd == 'play':
+          if not self._is_running():
+            logger.info(f'running: {self.vlc_args}')
+            self.proc = subprocess.Popen(args=self.vlc_args, preexec_fn=os.setpgrp)
+          f = open(self.command_file_path, 'w')
+          f.write('play')
+          f.close()
+          self.state = 'playing'
+
+        if cmd == 'next':
+          self.next_song()
+
+        if cmd == 'prev':
+          # Restart current song if we are past the first 3 seconds, otherwise go back
+
+          if self._ended_timeout > datetime.datetime.now():
+            self.previous_song()
+          else:
+            self.change_song(self.song_index)
+
+  def info(self) -> models.SourceInfo:
+    self.supported_cmds = ['play', 'pause']
+    if self.song_index != 0 and self.playing in self.song_list:
+      self.supported_cmds.append('prev')
+    if self.song_index != len(self.song_list) - 1 and self.playing in self.song_list:
+      self.supported_cmds.append('next')
+
+    img = 'static/imgs/no_note.png'
+    if self.playing != '':
+      img = 'static/imgs/note.png'
+    source = models.SourceInfo(name=self.full_name(),
+                               state=self.state,
+                               img_url=img,
+                               supported_cmds=self.supported_cmds,
+                               type=self.stream_type)
+    src_config_folder = f"{utils.get_folder('config')}/srcs/v{self.src}"
+    loc = f'{src_config_folder}/currentSong'
+    try:
+      with open(loc, 'r', encoding='utf-8') as file:
+        data = json.loads(file.read())
+        source.track = self.playing.split('/')[-1]
+        self.ended = data['state'] == 'ENDED'
+    except Exception:
+      pass
+
+    if self.state == 'playing' and self.playing in self.song_list and self.ended and self.song_index < len(self.song_list) - 1 and self._ended_timeout < datetime.datetime.now():
+      self.next_song()
+    elif self.ended and (self.song_index >= len(self.song_list) - 1 or self.playing not in self.song_list) and self._ended_timeout < datetime.datetime.now():
+      self.state = 'paused'
+    return source
```

Comment:
Does this assume that this info method is polled at a specific interval? It is reasonably possible for this not to have any clients polling.